src/vec-dealloc.md

```diff
@@ -7,20 +7,17 @@ ask Rust if `T` `needs_drop` and omit the calls to `pop`. However in practice
 LLVM is *really* good at removing simple side-effect free code like this, so I
 wouldn't bother unless you notice it's not being stripped (in this case it is).
```

Comment:
Note that the above is only true if LLVM can prove that the loop is finite. This is true in the case of `Vec`, but for example LLVM has no way to know that a linked list has no cycles.

Reply:
Makes sense. And based on the docs for [needs_drop](https://doc.rust-lang.org/std/mem/fn.needs_drop.html) it sounds like even in the `Vec` case LLVM might not eliminate the loop in debug builds. In any case, it is probably fragile to rely on LLVM for this behavior. I'm thinking of a few options for how we could address this:

1. Leave the code as is, but change the text to make the risk more clear.
2. Use a `mem::needs_drop` to skip the loop if `T` is `!Drop`.
3. Instead of an explicit loop, use `ptr::drop_in_place` (which is what the standard library `Vec` does).
4. Start with the current code, but within the section discuss its pitfalls and iterate on it, progressing to 2 and/or 3.

I would lean toward option 4, since currently this section is very short and could benefit from more discussion and examples to illustrate the usage of `mem::needs_drop` and `ptr::drop_in_place`. If we go with 2, 3, or 4, it will also require changes to `vec-raw.md` and `vec-final.md` for consistency. If you don't mind, do you think I could create a separate PR for this change, since it addresses a separate issue from the current PR?

Reply:
I agree that option 4 sounds like the best way to go. It's fine to do this in a separate PR.